lib/search_parameter_parser.rb

```
@@ -251,21 +263,117 @@ def return_fields
   end
 
   def filters
-    filters = {}
-    @params.each do |key, values|
-      if (m = key.match(/\Afilter_(.*)/))
-        field = m[1]
-        if ALLOWED_FILTER_FIELDS.include? field
-          filters[filter_name_lookup(field)] = values
+    filters = []
+    @params.each do |filter_param, values|
+      if filter_param.start_with?("filter_")
+        field = filter_param.sub("filter_", "")
+        if allowed_filter_fields.include? field
+          filters.push(build_filter(filter_name_lookup(field), values))
         else
           @errors << %{"#{field}" is not a valid filter field}
         end
-        @used_params << key
+        @used_params << filter_param
       end
     end
     filters
   end
 
+
+  class Filter
+    attr_reader :field_name, :values
+
+    def initialize(field_name, values)
+      @field_name = field_name
+      @values = Array(values)
+    end
+
+    def type
+      raise NotImplementedError
+    end
+
+    def ==(other)
+      [field_name, values] == [other.field_name, other.values]
+    end
+  end
+
+  class DateFieldFilter < Filter
+    def initialize(field_name, values)
+      super
+      @values = parse_dates(@values)
+    end
+
+    def type
+      "date"
+    end
+
+    private
+
+    def parse_dates(values)
+      values.map { |combined_from_and_to|
+        from = parse_date(combined_from_and_to.match(/from:([^,]+)/)[1])
+        to = parse_date(combined_from_and_to.match(/to:([^,]+)/)[1])
+
+        Value.new(from, to)
+      }
+    end
+
+    def parse_date(string)
+      Date.parse(string) rescue nil
```

If I'm reading this right, this swallows any syntax errors in submitted dates silently.  Let's not do that, please.  Other invalid inputs result in a 422 HTTP status with an explanatory message.